### ISSUE TITLE
Use rustls as an TLS engine

### ISSUE THREAD

**Issue by sagebind** [sagebind/isahc#199]
Offer [rustls](https://github.com/ctz/rustls) as an optional TLS engine directly. The TLS engine is decided by libcurl, so this is mostly something that would be work in the curl crate. This issue is just to track its availability in Isahc.

This is an alternative to #22. MesaLink's development seems to have stalled as of late and does not support ALPN, which is pretty much a blocker for making it a useful TLS engine for us (since it is required for HTTP/2).

**Comment by xy137** [comment]
has there been any progress on this?

**Comment by sagebind** [comment]
Nope, not yet, but there's an upstream issue to add support for it in the curl crate here: https://github.com/alexcrichton/curl-rust/issues/341

**Comment by XtremeDevX** [comment]
Would really love this - this would make for a great performance improvement!

**Comment by sagebind** [comment]
The first step for this has at long last landed, as the upstream curl crate now has a `rustls` crate feature: https://github.com/alexcrichton/curl-rust/issues/341.

Our work on Isahc's end is not finished though, as rustls does not use the operating system's trusted root certificates by default which is going to be an expected feature for Isahc (though potientially behind a separate crate feature). I'll keep the list of remaining tasks up-to-date in the issue description from here on out for more granular tracking.

In the meantime, you can now enable the `unstable-rustls-tls` crate feature on Isahc if you pull from the latest Git commit to start using rustls, though I expect there to be some rough edges at the moment.

**Comment by sagebind** [comment]
Adding "breaking" label to this, since the way Isahc 1.0 is configured, the native TLS engine is always enabled with no way of opting-out. We need to offer rustls and the native TLS engines as separate features that can be enabled or disabled, which is a breaking change.

2.0 is likely going to be a soon(ish) release anyway (a few months away probably) so seems like a good time to make rustls support part of that effort.

**Comment by 89z** [comment]
https://github.com/sagebind/isahc/issues/410

**Comment by seanpianka** [comment]
Hello, is there a tracking issue for the 2.0 release? I am looking forward to seeing this feature-flag stabilized. Thanks!

**Comment by sagebind** [comment]
There is no tracking issue, but there's a milestone here: https://github.com/sagebind/isahc/milestone/13. There is no due date for version 2.0, it'll be ready when it is ready.

**Comment by lcmgh** [comment]
Any update on this?

**Comment by sagebind** [comment]
@lcmgh Nope, sorry. I am currently taking a break from open-source work while I deal with some time-consuming projects in my personal life. It may be autumn this year before I can resume working on this. But thanks for your interest! I still plan on pushing this to the finish line, despite delays.

**Comment by sagebind** [comment]
This has been implemented in #491 and will be available in the v2 release in just a week or so.

**Comment by sagebind** [comment]
[Isahc 2.0 has now been released](https://github.com/sagebind/isahc/releases/tag/isahc-v2.0.0) with rustls support. Enjoy!
